### PR TITLE
Add orderAsc() and orderDesc() to solve sorting on expressions

### PR DESCRIPTION
**Pull request by markstory**
The current order() method has a limitation which makes it impossible to use expression objects that also have direction. These new methods makeit easier to support directed expression objects.

I've introduced a new expression object as I felt keeping OrderByExpression simple was worth the extra bit of cost. Additionally this makes SQL injection in the ORM internals more difficult as we rely on the callers to provide a safe expression. I also experimented with an array format, but felt it left us vulnerable to the issues we've had in the past around request data manipulation.

Refs #7163

### CODE REVIEW

**Comment by markstory** [comment]
I'm happy to put this into 3.1 if people feel that is a better place for it.

File: src/Database/Expression/OrderClauseExpression.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+use Cake\Database\ExpressionInterface;
+use Cake\Database\ValueBinder;
+
+/**
+ * An expression object for complex ORDER BY clauses
+ *
+ * @internal
+ */
+class OrderClauseExpression implements ExpressionInterface
```
Comment (lorenzo) [inline src/Database/Expression/OrderClauseExpression.php]: I think this needs to implement FieldInterface too

**Comment by lorenzo** [comment]
I fixed a variable name error in the code

File: src/Database/Expression/OrderClauseExpression.php (hunk 1 above)
Comment (markstory) [inline src/Database/Expression/OrderClauseExpression.php]: I'll do that and add tests for quoting on the generated queries.

Comment (markstory) [inline src/Database/Expression/OrderClauseExpression.php]: Done.

**Comment by markstory** [comment]
Tests are passing now.

**Comment by lorenzo** [comment]
Great job!

**Comment by ionas** [comment]
Thank you!

File: src/Database/Query.php [hunk 2]
```diff
@@ -941,13 +945,65 @@ public function order($fields, $overwrite = false)
...
+
+        if (!$this->_parts['order']) {
+            $this->_parts['order'] = new OrderByExpression();
+        }
+        $this->_parts['order']->add(new OrderClauseExpression($field, 'ASC'));
+        return $this;
+    }
+
+    /**
+     * Add an ORDER BY clause with an ASC direction.
```
Comment (LACabeza) [inline src/Database/Query.php]: This comment should be DESC direction.

Comment (markstory) [inline src/Database/Query.php]: This comment would be best as a pull request.

**Pull request merged**